crates/worker/src/partition/snapshots/repository.rs

```diff
@@ -0,0 +1,262 @@
+                .0
+                .into()
+        };
+
+        let staging_path = base_dir.join("snapshot-staging");
+        tokio::fs::create_dir_all(&staging_path).await?;
+
+        // prefix must be stripped of any leading slash and, unless zero-length, end in a single "/" character
+        let prefix: String = destination.path().into();
+        let prefix = match prefix.as_str() {
+            "" | "/" => "".to_string(),
+            prefix => format!("{}/", prefix.trim_start_matches('/').trim_end_matches('/')),
+        };
+
+        Ok(SnapshotRepository {
+            object_store,
+            destination,
+            prefix,
+            staging_path,
+        })
+    }
+
+    /// Write a partition snapshot to the snapshot repository.
+    pub(crate) async fn put(
+        &self,
+        metadata: &PartitionSnapshotMetadata,
+        local_snapshot_path: PathBuf,
+    ) -> anyhow::Result<()> {
+        let snapshot_id = metadata.snapshot_id;
+        let partition_id = metadata.partition_id;
+        let lsn = metadata.min_applied_lsn;
+
+        debug!(
+            %lsn,
+            "Publishing partition snapshot to: {}",
+            self.destination,
+        );
+
+        // All common object stores list objects in lexicographical order, with no option for
+        // reverse order. We inject an explicit sort key into the snapshot prefix to make sure that
+        // the latest snapshot is always first.
+        let inverted_sort_key = format!("{:016x}", u64::MAX - lsn.as_u64());
+
+        let metadata_json_path = local_snapshot_path.join("metadata.json");
+        let metadata_key = format!(
+            "{prefix}metadata/{partition_id}/{sk}_{snapshot_id}-lsn_{lsn}.json",
+            prefix = self.prefix,
+            sk = inverted_sort_key,
+            lsn = metadata.min_applied_lsn,
+        );
+
+        let snapshot_archive_path = self.staging_path.join(format!("{snapshot_id}.tar"));
+        let snapshot_tarball = tokio::fs::File::create_new(snapshot_archive_path.as_path()).await?;
+        let snapshot_key = format!(
+            "{prefix}snapshot/{partition_id}/{sk}_{snapshot_id}-lsn_{lsn}.tar",
+            prefix = self.prefix,
+            sk = inverted_sort_key,
+            lsn = metadata.min_applied_lsn,
+        );
+
+        let mut tar = tokio_tar::Builder::new(snapshot_tarball);
+        debug!(
+            "Creating snapshot tarball from {:?} as {:?}",
+            local_snapshot_path.as_path(),
+            tar.get_ref()
+        );
+        tar.append_dir_all(".", local_snapshot_path.as_path())
+            .await?;
+        tar.finish().await?;
+
+        let key = object_store::path::Path::from(snapshot_key.as_str());
+        let put_result =
+            put_snapshot_object(snapshot_archive_path.as_path(), &key, &self.object_store).await?;
+        debug!(
+            etag = put_result.e_tag.unwrap_or_default(),
+            "Successfully published snapshot to: {snapshot_key}",
+        );
```

Comment:
Writing the tarball to local disk is silly: we incur a lot of costs in the process in the form of IO and disk space usage. I initially tried to use `tokio::io::simplex` but backed out after I realized I need to implement framing on top of the stream, as dropping the writer end doesn't signal EOF to the reader (https://github.com/tokio-rs/tokio/issues/6914). Is it worth continuing in this direction, or should we back out and just upload individual SSTs?